### PR TITLE
[WIP] Refactor signing of receive address

### PR DESCRIPTION
**Pull request by arnabCypherock**

### CODE REVIEW

**Comment by vbcypherock** [comment]
Task linked: [CHI-2055 Changelly Hardware Enabler](https://app.clickup.com/t/37308523/CHI-2055)

**Review by ujjwal-cyph — CHANGES_REQUESTED** [review]

File: common/libraries/util/utils.h [hunk 1]
```diff
@@ -352,3 +352,22 @@ bool convert_byte_array_to_decimal_string(const uint8_t len,
                                           char *amount_decimal_string,
                                           const size_t amount_decimal_string_size);
 #endif
+
+#include "atca_host.h"
+#include "controller_level_four.h" // for attecc_slot_define_t
```
Comment (ujjwal-cyph) [inline common/libraries/util/utils.h:357]: do not import this header. instead move the definition here.

File: src/level_four/core/controller/receive_transaction_controller.c [hunk 2]
```diff
@@ -65,9 +65,20 @@
 #include "controller_tap_cards.h"
 #include "shamir_wrapper.h"
 
+#include "atca_host.h"
+
+#define AUTH_DATA_SERIAL_SIGN_MSG_SIZE      (POSTFIX1_SIZE + POSTFIX2_SIZE + SIGNATURE_SIZE + DEVICE_SERIAL_SIZE)
+#define AUTH_DATA_CHALLENGE_SIGN_MSG_SIZE   (POSTFIX1_SIZE + POSTFIX2_SIZE + SIGNATURE_SIZE)
```
Comment (ujjwal-cyph) [inline src/level_four/core/controller/receive_transaction_controller.c:71]: move this to utils

File: src/level_four/core/controller/receive_transaction_controller.c [hunk 3]
```diff
@@ -187,10 +198,88 @@ void receive_transaction_controller()
     } break;
 
     case RECV_TXN_DISPLAY_ADDR: {
-        uint8_t data[1 + sizeof(receive_transaction_data.address)];
-        data[0] = 1;  // confirmation byte
-        memcpy(data + 1, receive_transaction_data.address, sizeof(receive_transaction_data.address));
-        transmit_data_to_app(RECV_TXN_USER_VERIFIED_ADDRESS, data, strlen((const char *)data));
+        if(receive_transaction_auth_data.is_auth) {
+            uint8_t payload[43 + SESSION_ID_SIZE];
```
Comment (ujjwal-cyph) [inline src/level_four/core/controller/receive_transaction_controller.c:202]: keep the signing inside a function in utils

File: src/level_four/core/controller/receive_transaction_controller.c [hunk 4]
```diff
@@ -187,10 +198,88 @@ void receive_transaction_controller()
...
+            uint8_t payload_size = 0;
+            memcpy(payload, receive_transaction_data.address, sizeof(receive_transaction_data.address));
+            memcpy(payload + payload_size, receive_transaction_auth_data.session_id, sizeof(receive_transaction_auth_data.session_id));
+            payload_size += sizeof(receive_transaction_auth_data.session_id);
+
+            uint8_t hash[32];
```
Comment (ujjwal-cyph) [inline src/level_four/core/controller/receive_transaction_controller.c:208]:
```suggestion
            uint8_t hash[32] = {0};
```

File: common/coin_support/coin_utils.c [hunk 5]
```diff
@@ -168,6 +168,34 @@ int64_t byte_array_to_recv_txn_data(Receive_Transaction_Data *txn_data_ptr,const
...
+    
+    s_memcpy(&(txn_data_auth_ptr->is_auth), data_byte_array, size, sizeof(txn_data_auth_ptr->is_auth), &offset);
+    s_memcpy(txn_data_auth_ptr->session_id, data_byte_array, size, sizeof(txn_data_auth_ptr->session_id), &offset);
+    
+    return offset;
+}
+
+int64_t signed_recv_addr_to_byte_array(const Signed_Receive_Address *signed_recv_addr_ptr, uint8_t *signed_recv_addr_byte_array) {
+        int64_t offset = 0;
+        memcpy(signed_recv_addr_byte_array, signed_recv_addr_ptr->session_id, sizeof(signed_recv_addr_ptr->session_id));
```
Comment (ujjwal-cyph) [inline common/coin_support/coin_utils.c:183]: put an overflow check on `signed_recv_addr_byte_array`

File: common/libraries/util/utils.c [hunk 6]
```diff
@@ -478,4 +482,144 @@ bool convert_byte_array_to_decimal_string(const uint8_t len,const uint8_t decima
 
     LOG_INFO("amt %s %d:%d", amount_string, decimal, i);
     return true;
+}
+
+
+void helper_get_gendig_hash(atecc_slot_define_t slot, uint8_t *data, uint8_t *digest, uint8_t *postfix)
```
Comment (ujjwal-cyph) [inline common/libraries/util/utils.c:488]: keep all additions in a new file `atecc_utils`

**Review by ujjwal-cyph — CHANGES_REQUESTED** [review]

**Comment by ujjwal-cyph** [comment]
replaced by https://github.com/Cypherock/x1_wallet_firmware/pull/69

**Pull request closed without merging**